File: android/src/com/dsh/calc/SciFunc.java
package com.dsh.calc;

import java.math.BigInteger;
import java.util.Locale;

/**
 * 科学计算函数扩展(与桌面版 scifunc.cpp 语义一致)。
 * 三角/对数/指数函数走 double 近似, 结果可靠位数约 15 位;
 * abs/floor/ceil/round/fact 走精确大数路径。
 */
public final class SciFunc {

    private SciFunc() {}

    // ======================= 精确函数 =======================

    public static BigDec abs(BigDec x) {
        if (x.isZero() || !x.neg) return x;
        BigDec r = x.copy();
        r.neg = false;
        r.approx = false;
        r.reliable = -1;
        return r;
    }

    public static BigDec floor(BigDec x) {
        if (x.isZero()) return BigDec.of(0);
        if (x.e10 >= 0) return copyClean(x);
        BigInteger p = BigDec.TEN.pow((int) (-x.e10));
        BigInteger[] qr = x.mant.divideAndRemainder(p);
        BigInteger q = qr[0];
        if (x.neg && qr[1].signum() != 0) q = q.add(BigInteger.ONE);
        return makeInt(x.neg, q);
    }

    public static BigDec ceil(BigDec x) {
        if (x.isZero()) return BigDec.of(0);
        if (x.e10 >= 0) return copyClean(x);
        BigInteger p = BigDec.TEN.pow((int) (-x.e10));
        BigInteger[] qr = x.mant.divideAndRemainder(p);
        BigInteger q = qr[0];
        if (!x.neg && qr[1].signum() != 0) q = q.add(BigInteger.ONE);
        return makeInt(x.neg, q);
    }

    public static BigDec round(BigDec x) {
        if (x.isZero()) return BigDec.of(0);
        if (x.e10 >= 0) return copyClean(x);
        BigInteger p = BigDec.TEN.pow((int) (-x.e10));
        BigInteger[] qr = x.mant.divideAndRemainder(p);
        BigInteger q = qr[0];
        if (qr[1].shiftLeft(1).compareTo(p) >= 0) q = q.add(BigInteger.ONE);
        return makeInt(x.neg, q);
    }

    public static BigDec fact(BigDec x, Limits L) {
        if (x.neg) throw new CalcException("阶乘仅对非负整数有定义");
        Long n = toNonNegativeLong(x);
        if (n == null) throw new CalcException("阶乘参数过大或不是整数");
        if (n > 1) {
            // 和桌面版一致的位数预检, 避免真的去算一个天文数字
            double est = n * Math.log10(n / Math.E) + 0.5 * Math.log10(2.0 * Math.PI * n);
            if (!Double.isFinite(est) || est > 1.0e12)
                throw new CalcException("阶乘结果位数过多, 无法计算");
            BigDec estDec = new BigDec(false, BigInteger.ONE, (long) est);
            BigDec.checkSize(estDec, L, "阶乘");
        }
        BigInteger r = BigInteger.ONE;
        for (long i = 2; i <= n; ++i) r = r.multiply(BigInteger.valueOf(i));
        BigDec out = new BigDec(false, r, 0);
        out.approx = false;
        out.reliable = -1;
        BigDec.checkSize(out, L, "阶乘");
        return out;
    }

    // ======================= 角度转换 =======================

    public static BigDec degToRad(BigDec x, Limits L) {
        return approx(BigDec.toDouble(x) * Math.PI / 180.0, L);
    }

    public static BigDec radToDeg(BigDec x, Limits L) {
        return approx(BigDec.toDouble(x) * 180.0 / Math.PI, L);
    }

    // ======================= 三角函数(弧度制) =======================

    public static BigDec sin(BigDec x, Limits L) { return approx(Math.sin(BigDec.toDouble(x)), L); }
    public static BigDec cos(BigDec x, Limits L) { return approx(Math.cos(BigDec.toDouble(x)), L); }

    public static BigDec tan(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        double c = Math.cos(v);
        if (Math.abs(c) < 1e-18) throw new CalcException("tan 在该点无定义 (cos(x)=0)");
        double r = Math.tan(v);
        if (Double.isNaN(r) || Double.isInfinite(r))
            throw new CalcException("tan 结果超出可表示范围或无定义");
        return approx(r, L);
    }

    public static BigDec asin(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        if (Math.abs(v) > 1.0 + 1e-14) throw new CalcException("asin 定义域为 [-1, 1]");
        if (v > 1.0) v = 1.0;
        if (v < -1.0) v = -1.0;
        return approx(Math.asin(v), L);
    }

    public static BigDec acos(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        if (Math.abs(v) > 1.0 + 1e-14) throw new CalcException("acos 定义域为 [-1, 1]");
        if (v > 1.0) v = 1.0;
        if (v < -1.0) v = -1.0;
        return approx(Math.acos(v), L);
    }

    public static BigDec atan(BigDec x, Limits L) { return approx(Math.atan(BigDec.toDouble(x)), L); }

    // ======================= 双曲函数 =======================

    public static BigDec sinh(BigDec x, Limits L) { return approx(Math.sinh(BigDec.toDouble(x)), L); }
    public static BigDec cosh(BigDec x, Limits L) { return approx(Math.cosh(BigDec.toDouble(x)), L); }
    public static BigDec tanh(BigDec x, Limits L) { return approx(Math.tanh(BigDec.toDouble(x)), L); }

    // ======================= 指数与对数 =======================

    public static BigDec exp(BigDec x, Limits L) {
        double r = Math.exp(BigDec.toDouble(x));
        if (Double.isNaN(r) || Double.isInfinite(r))
            throw new CalcException("exp 结果超出可表示范围");
        return approx(r, L);
    }

    public static BigDec log(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        if (v <= 0) throw new CalcException("对数的真数必须大于 0");
        return approx(Math.log(v), L);
    }

    public static BigDec log10(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        if (v <= 0) throw new CalcException("对数的真数必须大于 0");
        return approx(Math.log10(v), L);
    }

    public static BigDec log2(BigDec x, Limits L) {
        double v = BigDec.toDouble(x);
        if (v <= 0) throw new CalcException("对数的真数必须大于 0");
        return approx(Math.log(v) / Math.log(2.0), L);
    }

    // ======================= 内部工具 =======================

    private static BigDec approx(double v, Limits L) {
        if (Double.isNaN(v) || Double.isInfinite(v))
            throw new CalcException("结果超出可表示范围或无定义");
        BigDec r = BigDec.fromDecimalString(String.format(Locale.US, "%.15g", v));
        r.approx = true;
        r.reliable = 15;
        return BigDec.roundSig(r, Math.min(L.precision, 15));
    }

    private static BigDec copyClean(BigDec x) {
        BigDec r = x.copy();
        r.approx = false;
        r.reliable = -1;
        return r;
    }

    private static BigDec makeInt(boolean neg, BigInteger mant) {
        BigDec r = new BigDec(neg, mant, 0);
        r.approx = false;
        r.reliable = -1;
        r.strip();
        return r;
    }

    private static Long toNonNegativeLong(BigDec x) {
        if (x.isZero()) return 0L;
        if (x.neg || x.e10 < 0) return null;
        String s = x.mant.toString();
        if (s.length() + x.e10 > 19) return null;
        StringBuilder sb = new StringBuilder(s);
        for (long i = 0; i < x.e10; ++i) sb.append('0');
        try {
            return Long.parseLong(sb.toString());
        } catch (NumberFormatException ex) {
            return null;
        }
    }
}

File: src/scifunc.cpp
// scifunc.cpp -- 科学计算函数扩展实现
#include "scifunc.h"

#include <cmath>
#include <cstdio>
#include <limits>

namespace calc {

namespace {

constexpr long double PI_L = 3.141592653589793238462643383279502884L;

// 把 BigDec 转成 long double 做近似计算。超大/超小值会安全地饱和到 inf/0。
long double decToLongDouble(const BigDec& x) {
    if (x.mant.isZero()) return 0.0L;
    size_t n = x.mant.d.size();
    size_t take = std::min<size_t>(3, n);
    long double t = 0;
    for (size_t i = 0; i < take; ++i)
        t = t * (long double)BIG_BASE + (long double)x.mant.d[n - 1 - i];
    long double ex = (long double)(n - take) * (long double)BIG_BASE_DIGITS + (long double)x.e10;
    if (ex > 6000.0L) return x.neg ? -std::numeric_limits<long double>::infinity()
                                   :  std::numeric_limits<long double>::infinity();
    if (ex < -6000.0L) return 0.0L;
    long double v = t * std::pow(10.0L, ex);
    return x.neg ? -v : v;
}

// 把 long double 结果转回 BigDec。由于来源是浮点, 结果天然最多约 15~17 位可靠数字。
BigDec fromLongDouble(long double v) {
    if (!std::isfinite((double)v)) throw CalcError("结果超出可表示范围或无定义");
    char buf[64];
    std::snprintf(buf, sizeof(buf), "%.17Lg", v);
    BigDec r = decFromString(std::string(buf));
    r.approx = true;
    r.reliable = 15;
    return r;
}

BigDec roundApprox(long double v, const Limits& L) {
    BigDec r = fromLongDouble(v);
    size_t prec = L.precision < 15 ? L.precision : 15;
    return decRoundSig(r, prec);
}

bool toU64NonNegative(const BigDec& x, unsigned long long& out) {
    if (x.neg) return false;
    if (x.mant.isZero()) { out = 0; return true; }
    if (x.e10 < 0) return false;
    std::string s = bigToString(x.mant);
    if (s.size() + (size_t)x.e10 > 20) return false;
    s.append((size_t)x.e10, '0');
    if (s.size() > 20) return false;
    if (s.size() == 20 && s.compare("18446744073709551615") > 0) return false;
    out = std::strtoull(s.c_str(), nullptr, 10);
    return true;
}

// 递归二分求 [a,b] 的乘积 (a*(a+1)*...*b), 比朴素循环快很多。
BigInt factRange(unsigned long long a, unsigned long long b) {
    if (a > b) return bigFromU64(1);
    if (a == b) return bigFromU64(a);
    unsigned long long m = a + (b - a) / 2;
    return bigMul(factRange(a, m), factRange(m + 1, b));
}

}  // namespace

// ============================ 精确函数 ============================

BigDec decAbs(const BigDec& x) {
    if (x.mant.isZero() || !x.neg) return x;
    BigDec r = x;
    r.neg = false;
    r.approx = false;
    r.reliable = -1;
    return r;
}

BigDec decFloor(const BigDec& x) {
    if (x.mant.isZero()) return decZero();
    if (x.e10 >= 0) {
        BigDec r = x;
        r.approx = false;
        r.reliable = -1;
        return r;
    }
    size_t k = (size_t)(-x.e10);
    BigInt p = bigPow10(k);
    BigInt q, rem;
    bigDivMod(x.mant, p, q, rem);
    if (x.neg && !rem.isZero()) q = bigAdd(q, bigFromU64(1));
    BigDec r;
    r.mant = q;
    r.e10 = 0;
    r.neg = x.neg;
    r.approx = false;
    r.reliable = -1;
    decStrip(r);
    return r;
}

BigDec decCeil(const BigDec& x) {
    if (x.mant.isZero()) return decZero();
    if (x.e10 >= 0) {
        BigDec r = x;
        r.approx = false;
        r.reliable = -1;
        return r;
    }
    size_t k = (size_t)(-x.e10);
    BigInt p = bigPow10(k);
    BigInt q, rem;
    bigDivMod(x.mant, p, q, rem);
    if (!x.neg && !rem.isZero()) q = bigAdd(q, bigFromU64(1));
    BigDec r;
    r.mant = q;
    r.e10 = 0;
    r.neg = x.neg;
    r.approx = false;
    r.reliable = -1;
    decStrip(r);
    return r;
}

BigDec decRound(const BigDec& x) {
    if (x.mant.isZero()) return decZero();
    if (x.e10 >= 0) {
        BigDec r = x;
        r.approx = false;
        r.reliable = -1;
        return r;
    }
    size_t k = (size_t)(-x.e10);
    BigInt p = bigPow10(k);
    BigInt q, rem;
    bigDivMod(x.mant, p, q, rem);
    BigInt twice = bigMulSmall(rem, 2);
    if (bigCmp(twice, p) >= 0) q = bigAdd(q, bigFromU64(1));
    BigDec r;
    r.mant = q;
    r.e10 = 0;
    r.neg = x.neg;
    r.approx = false;
    r.reliable = -1;
    decStrip(r);
    return r;
}

BigDec decFact(const BigDec& x, const Limits& L) {
    if (x.neg) throw CalcError("阶乘仅对非负整数有定义");
    unsigned long long n = 0;
    if (!toU64NonNegative(x, n))
        throw CalcError("阶乘参数过大或不是整数");

    // 用 Stirling 公式估算位数, 避免真的去算一个天文数字。
    if (n > 1) {
        long double est = (long double)n * std::log10((long double)n / std::exp(1.0L))
                        + 0.5L * std::log10(2.0L * PI_L * (long double)n);
        if (!std::isfinite((double)est) || est > 1.0e12L)
            throw CalcError("阶乘结果位数过多, 无法计算");
        BigDec estDec;
        estDec.mant = bigFromU64(1);
        estDec.e10 = (long long)est;
        decCheckSize(estDec, L, "阶乘");
    }

    BigDec r;
    if (n <= 1) {
        r = decFromU64(1);
    } else {
        r.mant = factRange(2, n);
        r.e10 = 0;
    }
    r.approx = false;
    r.reliable = -1;
    decCheckSize(r, L, "阶乘");
    return r;
}

// ============================ 角度转换 ============================

BigDec degToRad(const BigDec& x, const Limits& L) {
    return roundApprox(decToLongDouble(x) * PI_L / 180.0L, L);
}

BigDec radToDeg(const BigDec& x, const Limits& L) {
    return roundApprox(decToLongDouble(x) * 180.0L / PI_L, L);
}

// ============================ 三角函数 (弧度制) ============================

BigDec decSin(const BigDec& x, const Limits& L) {
    return roundApprox(std::sin(decToLongDouble(x)), L);
}

BigDec decCos(const BigDec& x, const Limits& L) {
    return roundApprox(std::cos(decToLongDouble(x)), L);
}

BigDec decTan(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    long double c = std::cos(v);
    if (std::fabsl(c) < 1e-18L)
        throw CalcError("tan 在该点无定义 (cos(x)=0)");
    long double r = std::tan(v);
    if (!std::isfinite((double)r))
        throw CalcError("tan 结果超出可表示范围或无定义");
    return roundApprox(r, L);
}

BigDec decAsin(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    if (std::fabsl(v) > 1.0L + 1e-14L)
        throw CalcError("asin 定义域为 [-1, 1]");
    if (v > 1.0L) v = 1.0L;
    if (v < -1.0L) v = -1.0L;
    return roundApprox(std::asin(v), L);
}

BigDec decAcos(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    if (std::fabsl(v) > 1.0L + 1e-14L)
        throw CalcError("acos 定义域为 [-1, 1]");
    if (v > 1.0L) v = 1.0L;
    if (v < -1.0L) v = -1.0L;
    return roundApprox(std::acos(v), L);
}

BigDec decAtan(const BigDec& x, const Limits& L) {
    return roundApprox(std::atan(decToLongDouble(x)), L);
}

// ============================ 双曲函数 ============================

BigDec decSinh(const BigDec& x, const Limits& L) {
    return roundApprox(std::sinh(decToLongDouble(x)), L);
}

BigDec decCosh(const BigDec& x, const Limits& L) {
    return roundApprox(std::cosh(decToLongDouble(x)), L);
}

BigDec decTanh(const BigDec& x, const Limits& L) {
    return roundApprox(std::tanh(decToLongDouble(x)), L);
}

// ============================ 指数与对数 ============================

BigDec decExp(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    long double r = std::exp(v);
    if (!std::isfinite((double)r))
        throw CalcError("exp 结果超出可表示范围");
    return roundApprox(r, L);
}

BigDec decLog(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    if (v <= 0.0L) throw CalcError("对数的真数必须大于 0");
    return roundApprox(std::log(v), L);
}

BigDec decLog10(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    if (v <= 0.0L) throw CalcError("对数的真数必须大于 0");
    return roundApprox(std::log10(v), L);
}

BigDec decLog2(const BigDec& x, const Limits& L) {
    long double v = decToLongDouble(x);
    if (v <= 0.0L) throw CalcError("对数的真数必须大于 0");
    return roundApprox(std::log(v) / std::log(2.0L), L);
}

}  // namespace calc
